apps/sms/services/js/messaging/messaging_server.js

```diff
@@ -0,0 +1,542 @@
+  },
+
+  markMessagesRead: function ms_markMessagesRead(list, isRead = true) {
+    if (!mozMobileMessage || !list.length) {
+      return;
+    }
+
+    // We chain the calls to the API in a way that we make no call to
+    // 'markMessageRead' until a previous call is completed. This way any
+    // other potential call to the API, like the one for getting a message
+    // list, could be done within the calls to mark the messages as read.
+
+    var id = list.pop();
+    // TODO: Third parameter of markMessageRead is return read request.
+    //       Here we always return read request for now, but we can let user
+    //       decide to return request or not in Bug 971658.
+    var req = mozMobileMessage.markMessageRead(id, isRead, true);
+    // isRead == false i.e mark as unread case, marking only one message
+    // as unread is sufficient.
+    req.onsuccess = (function onsuccess() {
+      if (!list.length || !isRead) {
+        return;
+      } else if (isRead) {
+        this.markMessagesRead(list, isRead);
+      }
+    }).bind(this);
+
+    req.onerror = function onerror() {
+      console.error(
+        'Error while marking message %d as read: %s', id, this.error.name
+      );
+    };
+  },
+
+  getSegmentInfo: function ms_getSegmentInfo(text) {
+    if (!(mozMobileMessage && mozMobileMessage.getSegmentInfoForText)) {
+      return Promise.reject(new Error('mozMobileMessage is unavailable.'));
+    }
+
+    var defer = Utils.Promise.defer();
+
+    var request = mozMobileMessage.getSegmentInfoForText(text);
+    request.onsuccess = function onsuccess(e) {
+      defer.resolve(e.target.result);
+    };
+
+    request.onerror = function onerror(e) {
+      defer.reject(e.target.error);
+    };
+
+    return defer.promise;
+  },
+
+  /**
+   * findThreadFromNumber
+   *
+   * Find a SMS/MMS thread from a number.
+   * @return Promise that resolve to a threadId or rejected if not found
+   */
+  findThreadFromNumber: function ms_findThread(number) {
+
+    function checkCandidate(message) {
+      var isMessageInThread =
+        isMessageBelongTo1to1Conversation(number, message);
+
+      if (isMessageInThread) {
+        threadId = message.threadId;
+        // we need to set the current threadId,
+        // because we start sms app in a new window
+        Threads.registerMessage(message);
+        return false; // found the message, stop iterating
+      }
+    }
+
+    var threadId = null;
+    var deferred = Utils.Promise.defer();
+
+    MessagingService.getMessages({
+      filter: { numbers: [number] },
+      each: checkCandidate,
+      done: function() {
+        if (threadId == null) {
+          deferred.reject(new Error('No thread found for number: ' + number));
+          return;
+        } else {
+          deferred.resolve(threadId);
+          return;
+        }
+      }
+    });
+
+    return deferred.promise;
+  }
+};
+
+MessagingService.initEventListener();
+
```

Comment:
I didn't encapsulate the MessagingService further because it didn't export to anywhere. But it's good to know if you have any opinion about the MessagingService.